Scripts to package, install, upgrade, run Mathesar

pavish opened the pull request:
Fixes #4304 
Fixes #4305 
Fixes #4306 
Fixes #4309 
Related to #4307 
Related to #4308 

This PR introduces three major scripts:
- `install.sh`: For installing Mathesar, invoked by the user directly. 
- `package.sh`: Called by our CI during releases, as well for development & for testing.
- `mathesar`: Script will be installed on the user's machine. Handles running mathesar.

The scripts should be mostly self-explanatory, there are comments where needed.

This PR also updates the GH workflow for updating releases with the packaged archive as a release asset. It adds one more that packages mathesar and sets it as a workflow artifact for testing purposes.

## For reviewers

### Documentation

- [Rendered user documentation](https://github.com/mathesar-foundation/mathesar/blob/simplify-install/docs/docs/administration/install-from-scratch.md).

Some elements in the documentation would not appear as it would on our docs site. If you're reviewing the documentation, build the MkDocs site locally off this branch.

> [!NOTE]
>
> Edit from Sean: ~~The link above ☝️ points to a specific commit which is now outdated.~~ I recommend you read the rendered user documentation by building the MkDocs site locally off this branch.

### Testing installation
- Follow the documentation, except use testing links.
- Move into the folder you'd like to install.
- Download the install.sh script from this branch
```
curl -sSfL https://github.com/mathesar-foundation/mathesar/releases/download/draft_TEST_ONLY_simplify-install/install.sh -o install.sh
chmod +x install.sh
```
- Install mathesar, specify the test package location with the one in this [testing release](https://github.com/mathesar-foundation/mathesar/releases/tag/draft_TEST_ONLY_simplify-install).
  - Provide a db connection string for the new django database. If not specified, the script will prompt you for it, it won't fail.
```
./install.sh . -c "postgres://mathesar:mathesar@localhost:5432/mathesar_django" --test-package-location "https://github.com/mathesar-foundation/mathesar/releases/download/draft_TEST_ONLY_simplify-install/mathesar.tar.gz"
```
- Note: The `--test-package-location` option is only needed here because we're asking the install script to target a specific link instead of the default.
- Open a new shell
- Run mathesar
```
mathesar run -p 9000
```
- Note: If you're attempting to install this on your dev environment (over docker), your existing dev environment variables will interfere with the installation. Use a different container or a separate environment.

## Work done in this PR
- [x] Write package script that packages required files/directories, builds frontend, compiles translations
- [x] Write install script that downloads package, installs python, setsup venv, installs requirements
- [x] Ask for postgres connection env variables from user while installing
- [x] Automatically setup secret key, do not ask for/replace existing env variables
- [x] Check for existing packages and uv installations before downloading
- [x] Make install script result in idempotent state
- [x] Check for existing venv's, python versions
- [x] Identify managed & unmanaged python instances, handle scenarios accordingly
- [x] Upload packages as GH artifacts
- [x] Update create-draft-release workflow to upload the package instead of frontend static_files 
- [x] Allow specifying directory while running install script
- [x] Allow specifying package location while running install script for testing purposes
- [x] Provide `--connection-string`, `--force-download-python`, `--no-prompt` options
		- `--no-prompt` would not ask the user for env vars (they need to be passed in)
		- `--no-prompt` would not repeat asking if connection string if invalid
		- `--no-prompt` cannot be used without `--connection-string`
- [x] Verify connection string when setting env vars (get connection string instead of all vars)
- [x] Write mathesar.sh and mathesar.dev.sh scripts, remove all run.sh scripts
- [x] Handle automatically setting environment variable from .env file when not passed
- [x] Set `mathesar` in PATH during installation
- [x] Add unit test for install.sh
- [x] Handle mathesar version checks during installation
- [x] Update user documentation
- [x] Add workflow to test installing Mathesar package

## Out of scope for this PR - Work to be done in future PRs
- Update release process wiki documentation
- Pull in uv license while packaging (or directly in repo) (More thought needed)
- Add command to mathesar script to one step install mathesar systemd service?
- Change package.sh to ignore patterns instead of specifying directories
		- Have a prodignore file?
		- (More thought needed)
- Add python tests for env variable setter
- ~~Add option to reset admin password to mathesar script? (More thought needed)~~


## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. --> 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `develop` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

pavish commented:
@mathemancer Please review code, documentation, and verify all changes in the PR.
@zackkrida @seancolsen Please review the documentation and test installation. You're invited to review code as well, if you're interested.
@Anish9901 I'm inviting you to test installation, if you're interested.

mathemancer commented on scripts/install.sh:
```diff
@@ -0,0 +1,772 @@
+}
+
+# For commands that should ideally never fail
+ensure() {
+  if ! "$@"; then err "Command failed: $*"; fi
+}
```
Same as above.

pavish replied:
See comment here: https://github.com/mathesar-foundation/mathesar/pull/4399#discussion_r2077379112

mathemancer commented on scripts/install.sh:
```diff
@@ -0,0 +1,772 @@
+    info "Downloading from $1 (using wget)..."
+    run_cmd wget "$1" -O "$2"
+  else
+    require_command "curl or wget"
```
I'd be shocked if a user doesn't have `wget` on a functioning *nix system. Also, wrapping in `run_cmd` here will distract in error cases (as above).

pavish replied:
See comment on error cases here: https://github.com/mathesar-foundation/mathesar/pull/4399#discussion_r2077379112

mathemancer commented on scripts/install.sh:
```diff
@@ -0,0 +1,772 @@
+
+#=======CONFIGURATIONS=========================================================
+
+MATHESAR_VERSION="0.2.3"
```
I don't like maintaining this in multiple places. Can it be injected during the build?

pavish replied:
Agreed, I don't like it too. I wanted to do this incrementally, but I'll try to get the changes for it done in this PR.

pavish replied:
I've made the necessary changes:
- mathesar script: Takes the version from mathesar/__init__.py
- install script: Gets injected during build

pavish commented on bin/mathesar:
```diff
@@ -0,0 +1,440 @@
+
+validate_port() {
+  if ! [[ "${MATHESAR_PORT}" =~ ^[0-9]+$ ]]; then
+    err "Invalid port: ${MATHESAR_PORT}. Port must be a positive integer."
```
I've removed this in https://github.com/mathesar-foundation/mathesar/pull/4399/commits/7e54cc58e7d3a1176cffc25887dae581dfa2aafa

pavish commented on bin/mathesar:
```diff
@@ -0,0 +1,440 @@
+    require_command grep
+    # netstat output may include IPv4 or IPv6 addresses. We match common delimiters.
+    if netstat -an | grep -E 'LISTEN' | grep -q "[.:]${MATHESAR_PORT}[[:space:]]"; then
+      err "Port ${MATHESAR_PORT} is already in use (checked with netstat). Please choose a different port."
```
This is removed in https://github.com/mathesar-foundation/mathesar/pull/4399/commits/7e54cc58e7d3a1176cffc25887dae581dfa2aafa

pavish commented on bin/mathesar:
```diff
@@ -0,0 +1,440 @@
+      err "Port ${MATHESAR_PORT} is already in use (checked with lsof). Please choose a different port."
+    fi
+  elif command_exists netstat; then
+    require_command grep
```
Removed in https://github.com/mathesar-foundation/mathesar/pull/4399/commits/7e54cc58e7d3a1176cffc25887dae581dfa2aafa

pavish commented on bin/mathesar:
```diff
@@ -0,0 +1,440 @@
+  if ! command_exists "$1"; then
+    err "Mathesar requires '$1' (command not found)"
+  fi
+}
```
> This ruins (or at least modifies) the user's expected error response when a command is expected, but missing.
> That error response is generally more useful than the replacement error response here. 

The normal error response from the shell is the same as the error response here.
Executing `echo "  asdf" | xadrgs`:
- zsh shows this: `zsh: command not found: xadrgs`
- bash shows this: `bash: xardgs: command not found`

> So, what value is this command bringing to offset that cost?

This check informs users of required commands at the very beginning of the script. This is particularly useful while installation, because the script notifies the user of all the required commands before beginning any work, rather than proceeding with the installation and failing due to a missing command somewhere mid-script.

I have however, removed the need for require_command in the mathesar startup script, because it only checks for xargs which is reasonable enough for us to expect it'd be present in the user's environment. If not present, I'm okay with the 'command not found' error thrown while attempting to run Mathesar. This is the commit: https://github.com/mathesar-foundation/mathesar/pull/4399/commits/7b66d1527c54f8d72c3468a143b7227bd7f288a4

pavish commented on bin/mathesar:
```diff
@@ -0,0 +1,440 @@
+        # If the option string includes any 'p' we now no longer allow clubbing.
+        shortopts="${1#-}"
+        if [[ "$shortopts" == *"p"* ]]; then
+          err "The -p option cannot be clubbed with other options."
```
I've made the error message more clear in https://github.com/mathesar-foundation/mathesar/pull/4399/commits/a71ebc5a5a5e7590ec97593c35510162320943b5.

pavish commented on bin/mathesar:
```diff
@@ -0,0 +1,440 @@
+# If no command is specified, print help and exit
+if [ $# -eq 0 ]; then
+  usage_msg
+  exit 0
```
I've changed it to `exit 2` in https://github.com/mathesar-foundation/mathesar/pull/4399/commits/cb5c126e9e065154c04b33e890218677ce58e146, which makes more sense as you mentioned.

I followed `docker` when writing this and it seems to return 0. But most tools return 2.

pavish commented on bin/mathesar:
```diff
@@ -0,0 +1,440 @@
+# For commands that should ideally never fail
+ensure() {
+  if ! "$@"; then err "Command failed: $*"; fi
+}
```
This is only used in the script in places where we expect the command to never fail.

The `ensure` function does the following:
- It shows the output from the executed commands.
   - This output might not always make it clear to the user that there was an error.
   - It also doesn't make it clear that this specific error needs user attention.
   - Eg., `cat: /dev/port: Permission denied` may be a blocking error for our use-case, or it could be a scenario where it's a nice-to-have like updating the bash profiles in which case it's okay if it fails, the user needn't have to worry.
- For the above reasons, it shows that the command failed (in red), informing the user that there was an error and it needs their attention.
- It shows that installation failed (in install script) and startup failed (in startup script).
- It exits the install/startup script with status 1, despite the status code from the failed command (to have more control over our exit codes).

I added the red color to prompt attention from the user. Capturing & coloring the stderr from the command would require us to execute the command in a subshell, which I'd like to avoid.

I made some changes in https://github.com/mathesar-foundation/mathesar/pull/4399/commits/6e989cf5aa7e000c230b0a265d1fd162366572f2 to make the message more useful. It now looks like this:

![Screenshot 2025-05-07 at 3 04 52 PM](https://github.com/user-attachments/assets/5eb6abc3-28d7-4172-89be-dcc4d9b1caa5)

pavish commented on bin/mathesar:
```diff
@@ -0,0 +1,440 @@
+
+select_python() {
+  if [[ "${NO_VENV}" = true ]]; then
+    echo "python"
```
Updated to use `python3`

pavish commented on bin/mathesar:
```diff
@@ -0,0 +1,440 @@
+}
+
+# Determine the script's real path only if it's a symlink
+if [[ -L "$0" ]]; then
```
We're handling symlinks here because we add a symlink at `local/.bin/mathesar` to this script during installation, and operations in this script require the original script's path.

I don't think there would ever be a case for someone sourcing this script, as you mentioned. I'd like to avoid handling that scenario.

pavish commented on bin/mathesar:
```diff
@@ -0,0 +1,440 @@
+
+verify_docker_setup() {
+  if [[ -z "${MATHESAR_DOCKER_IMAGE}" ]]; then
+    err "Cannot verify environment as Mathesar's docker image."
```
Updated in https://github.com/mathesar-foundation/mathesar/pull/4399/commits/78278bdc45331c0666a5ed184caa3cc4cde5b92e

pavish commented on bin/mathesar:
```diff
@@ -0,0 +1,440 @@
+        usage_err "Unknown option: $1"
+        ;;
+    esac
+  done
```
This is the first operation on the script and wrapping this in another function didn't seem to offer enough readability improvement. Also, whenever I read lengthy scripts, I try to find the area where the arguments are parsed and then read the script from there, I assumed others might want to do the same.

pavish commented:
@mathemancer This should be ready for a re-review.

> Figure out ways to reduce duplication between scripts, e.g., build them as part of the build process to add version info, util functions, etc.

I've handled versions. There are very few common util functions, and it would take more development effort/time and maintenance work injecting them during build than just letting them be, so I've not handled that.

> Reduce complexity as much as possible from each script. On this round, I noticed a lot of overdone (IMO) error handling that could just be removed.
> On the error handling front, please look for places where you can just lazily let things fail instead of trying to get out ahead of any error.

I've removed some error handling. But I do think the remaining part of the script are required, we can discuss after you take another look.

pavish commented:
@seancolsen 

> The guide seems to require administrators to supply their PostgreSQL credentials to install.sh via command line argument. Some people might not want to pass secrets on the command line like that. Could we prompt the user for the Postgres credentials if not passed? Maybe it already does this — I've not looked. If so, then we ought to document it.

The script already prompts the user during installation if the credentials are not supplied via the argument.

Agreed that we ought to document it. I'd like to handle docs updates in a separate PR once this one is merged.

> I've read over the new docs content and pushed https://github.com/mathesar-foundation/mathesar/commit/3ad7befef56d22d45d34bf09ed15dbbf85176d44 with some light cleanup, mostly to address structure, typos, markdown cleanliness, etc.

I have some concerns with the changes, however I think we can handle that separate from this PR. I added docs here so the reviewers can test the PR. I am mostly not comfortable with the term "Install from scratch", since we now provide static assets and the installation doesn't require as much as debugging as it used to.

seancolsen commented:
> I am mostly not comfortable with the term "Install from scratch"

That's fine with me. In this PR the nav menu still [says](https://github.com/mathesar-foundation/mathesar/blob/simplify-install/docs/mkdocs.yml#L11) "Install from scratch". So in my commit I was updating the page title to be more consistent with the nav menu (and simpler). I'm not attached to the "from scratch" naming. But I do want the title to be simple enough for readers to reason about as they compare our installation options. And I want the title in the page to be (roughly) consistent with the title in the nav menu.

mathemancer reviewed (commented):
This PR has over 1000 lines of shell script. That makes maintaining a positive first impression for users extremely difficult. It's too hard to validate and test, and is too dependent on local system utilities, etc. I _strongly_ suggest rewriting as much of each of these scripts in python (instead of shell) as possible. For example, there's currently a bug that could result in _corrupting a current Mathesar installation_ in the script, and I only found it by reading line-by-line, and having a pretty thorough knowledge of shell scripting. It would not surprise me (at all) if there were more such bugs.

For `install.sh`, you should just get `uv` working with minimal fuss or assumptions, then do everything else from the known python at that point.

Side note: Duplication between and within the scripts:
- output coloring (relevant since functions are duplicated that use it)
- `info`
- `err_msgonly`
- `command_exists`
- `run_err/err`
- `ensure`
- `mathesar-venv` directory name and location
- logic in `run_cmd` vs. `ensure`

*At this point in the review, I had a call with @pavish and discussed moving some logic to python. We agreed to try it, so I'll finish this review and start another tailored toward that goal.

pavish commented on scripts/install.sh:
```diff
@@ -0,0 +1,787 @@
+
+    info "Existing Mathesar detected in ${INSTALL_DIR} (version ${installed_ver})."
+    local cmp_result
+    cmp_result=$(compare_versions "${MATHESAR_VERSION}" "${installed_ver}")
```
Updated the function

pavish commented on scripts/install.sh:
```diff
@@ -0,0 +1,787 @@
+  if [[ "$greatest" == "$v1" ]]; then
+    return 1 # v1 newer
+  else
+    return -1 # v1 older
```
Updated

pavish commented:
@mathemancer This is ready for re-review.

I would like to add some unit tests for the shell scripts, however, I don't think that should block the review/merge of this PR. I can add that in a subsequent PR. This PR contains an integration test that runs as part of a GH workflow for the happy path scenario.

pavish commented on bin/mathesar_dev:
```diff
@@ -0,0 +1,21 @@
+
+BASE_DIR="$(cd "$(dirname "$0")/.." && pwd)"
+
+pushd "${BASE_DIR}" > /dev/null
```
Addressed

pavish commented on bin/mathesar_dev:
```diff
@@ -0,0 +1,21 @@
+  # client. Hence, instead of using this script, the web server
+  # can be directly started.
+  if [ -n "$NODE_MAJOR" ]; then
+      cd mathesar_ui
```
Addressed

pavish commented on bin/mathesar_dev:
```diff
@@ -0,0 +1,21 @@
+      # Run vite dev server only in dev mode
+      npm run dev &
+
+      cd ..
```
Addressed

pavish commented on bin/mathesar_dev:
```diff
@@ -0,0 +1,21 @@
+  python -m mathesar.install
+  python manage.py createsuperuser --no-input --username admin --email admin@example.com
+  python manage.py runserver 0.0.0.0:8000 && fg
+popd > /dev/null
```
Addressed

pavish commented on build-scripts/bin/mathesar.sh:
```diff
@@ -0,0 +1,395 @@
+  fi
+
+  info "Starting gunicorn webserver..."
+  exec "$gunicorn_bin" $gunicorn_opts
```
Addressed

pavish commented on build-scripts/bin/mathesar.sh:
```diff
@@ -0,0 +1,395 @@
+}
+
+print_mathesar_version() {
+  local version=$(get_mathesar_version)
```
Addressed

pavish commented on build-scripts/bin/mathesar.sh:
```diff
@@ -0,0 +1,395 @@
+#=======USAGE==================================================================
+
+usage_msg() {
+local version=$(get_mathesar_version)
```
Addressed

pavish commented on build-scripts/common_utilities.sh:
```diff
@@ -0,0 +1,81 @@
+  if [[ $status -ne 0 ]]; then
+    echo ""
+    danger "The command '$*' failed with status '$status' with the following error,${RESET}"
+    cat "$tmpfile" | (
```
Addressed

pavish commented on build-scripts/install/install.sh:
```diff
@@ -0,0 +1,422 @@
+    run_cmd "${UV_DIR}/uv" venv ./"${VENV_DIR_NAME}" --python "${PYTHON_SPEC}" --seed --relocatable
+
+    info "Activating Python virtual environment..."
+    source ./"${VENV_DIR_NAME}"/bin/activate
```
Addressed

pavish commented on build-scripts/install/install.sh:
```diff
@@ -0,0 +1,422 @@
+  local status=0
+
+  updated_content=$(
+    cat "${INSTALL_DIR}"/"${ENV_FILE_NAME}" | "${INSTALL_DIR}"/"${VENV_DIR_NAME}"/bin/python ./setup/process_env.py "$conn_str_argument"
```
Addressed

pavish commented on build-scripts/install/install.sh:
```diff
@@ -0,0 +1,422 @@
+      # repeatedly prompt the user for a valid PostgreSQL connection string.
+      while true; do
+        echo ""
+        read -ep "Enter PostgreSQL connection string (format: postgres://user:password@host:port/dbname): " conn_str
```
Addressed

pavish commented on build-scripts/install/install.sh:
```diff
@@ -0,0 +1,422 @@
+
+    info "Exporting environment variables to shell..."
+    set -a
+    source .env
```
Addressed

pavish commented on build-scripts/install/install_path_handler.sh:
```diff
@@ -0,0 +1,214 @@
+        return 1
+      fi
+    fi
+  elif ! . "${INSTALL_DIR}/bin/mathesar_path_source"; then
```
Addressed

pavish commented on build-scripts/install/install.sh:
```diff
@@ -0,0 +1,422 @@
+include_source "./install_path_handler.sh"
+
+MATHESAR_VERSION=___MATHESAR_VERSION___
+REQUIRED_UV_VERSION=___UV_VERSION___
```
Addressed

pavish commented:
@mathemancer Shellcheck is added to the GH workflow and linting errors are fixed.

mathemancer reviewed (commented):
Reviewing this much bash is slow going; sorry about that. I'm going to submit an in-progress review, since I'm stuck on actually trying the scripts. Please:

- document how to put the script in the required place for testing (or did I miss that)
- Address any preliminary comments.

I'll review again later this evening.

pavish commented:
The instructions in the PR description correspond to installing a package formed from a specific commit/release.
- Commit: https://github.com/mathesar-foundation/mathesar/pull/4399/commits/47a8fe190cec22960b2540154f95e60084744764
- Release: https://github.com/mathesar-foundation/mathesar/releases/tag/draft_TEST_ONLY_simplify-install

The packaging happens via a GH workflow.

---

**If you'd like to package Mathesar yourself locally, and then install your local package, follow the instructions below:**

NOTE: You only have to do this if you add more commits & the release no longer contains the latest changes, and you'd like to avoid creating a draft release for some reason.

*Instructions to package Mathesar:*

From within a dev environment, run:
```
python3 ./build-scripts/package/package.py
```

This would produce a `dist` directory at the root of the repo, with the files `install.sh` and `mathesar.tar.gz`.

*Instructions to install Mathesar from a local package:*

From within the dist folder, run:
```
chmod +x install.sh

./install.sh <install_dir> -c "postgres://testuser:testpass@localhost:5432/testdb" --test-package-location "file:////code/dist/mathesar.tar.gz"
```

Replace install_dir with the location you'd like to install Mathesar into.

mathemancer commented:
Okay, I tried the install with the updated script and package, but got the following error:

```
The command '/home/brent/test_installation/uv/uv pip install -r requirements.txt' failed with status '1' with the following error,                                                                                                                                                                                 
Using Python 3.13.3 environment at: mathesar-venv                                                                                                                                                                                                                                                                  
Resolved 30 packages in 2.03s                                                                                                                                                                                                                                                                                      
   Building drf-access-policy==1.5.0                                                                                                                                                                                                                                                                               
Downloading psycopg2-binary (2.9MiB)                                                                                                                                                                                                                                                                               
Downloading djangorestframework (1.0MiB)                                                                                                                                                                                                                                                                           
Downloading psycopg-binary (4.1MiB)                                                                                                                                                                                                                                                                                
 Downloading djangorestframework                                                                                                                                                                                                                                                                                   
Downloading django (7.6MiB)                                                                                                                                                                                                                                                                                        
Downloading cryptography (4.3MiB)                                                                                                                                                                                                                                                                                  
   Building clevercsv==0.8.2                                                                                                                                                                                                                                                                                       
 Downloading psycopg2-binary                                                                                                                                                                                                                                                                                       
 Downloading cryptography                                                                                                                                                                                                                                                                                          
 Downloading psycopg-binary                                                                                                                                                                                                                                                                                        
   Building sqlalchemy==1.4.54                                                                                                                                                                                                                                                                                     
      Built drf-access-policy==1.5.0                                                                                                                                                                                                                                                                               
 Downloading django                                                                                                                                                                                                                                                                                                
  × Failed to build `clevercsv==0.8.2`                                                                                                                                                                                                                                                                             
  ├─▶ The build backend returned an error                                                                                                                                                                                                                                                                          
  ╰─▶ Call to `setuptools.build_meta:__legacy__.build_wheel` failed (exit status: 1)                                                                                                                                                                                                                               
                                                                                                                                                                                                                                                                                                                   
      [stdout]                                                                                                                                                                                                                                                                                                     
      running bdist_wheel                                                                                                                                                                                                                                                                                          
      running build                                                                                                                                                                                                                                                                                                
      running build_py                                                                                                                                                                                                                                                                                             
      creating build/lib.linux-x86_64-cpython-313/clevercsv                                                                                                                                                                                                                                                        
      copying clevercsv/encoding.py -> build/lib.linux-x86_64-cpython-313/clevercsv                                                                                                                                                                                                                                
      copying clevercsv/__version__.py -> build/lib.linux-x86_64-cpython-313/clevercsv                                                                                                                                                                                                                             
      copying clevercsv/utils.py -> build/lib.linux-x86_64-cpython-313/clevercsv                                                                                                                                                                                                                                   
      copying clevercsv/_optional.py -> build/lib.linux-x86_64-cpython-313/clevercsv                                                                                                                                                                                                                               
      copying clevercsv/consistency.py -> build/lib.linux-x86_64-cpython-313/clevercsv                                                                                                                                                                                                                             
      copying clevercsv/_regexes.py -> build/lib.linux-x86_64-cpython-313/clevercsv                                                                                                                                                                                                                                
      copying clevercsv/escape.py -> build/lib.linux-x86_64-cpython-313/clevercsv                                                                                                                                                                                                                                  
      copying clevercsv/write.py -> build/lib.linux-x86_64-cpython-313/clevercsv                                                                                                                                                                                                                                   
      copying clevercsv/cparser_util.py -> build/lib.linux-x86_64-cpython-313/clevercsv                                                                                                                                                                                                                            
      copying clevercsv/wrappers.py -> build/lib.linux-x86_64-cpython-313/clevercsv                                                                                                                                                                                                                                
      copying clevercsv/_types.py -> build/lib.linux-x86_64-cpython-313/clevercsv                                                                                                                                                                                                                                  
      copying clevercsv/detect_type.py -> build/lib.linux-x86_64-cpython-313/clevercsv                                                                                                                                                                                                                             
      copying clevercsv/read.py -> build/lib.linux-x86_64-cpython-313/clevercsv                                                                                                                                                                                                                                    
      copying clevercsv/normal_form.py -> build/lib.linux-x86_64-cpython-313/clevercsv                                                                                                                                                                                                                             
      copying clevercsv/potential_dialects.py -> build/lib.linux-x86_64-cpython-313/clevercsv                                                                                                                                                                                                                      
      copying clevercsv/detect_pattern.py -> build/lib.linux-x86_64-cpython-313/clevercsv                                                                                                                                                                                                                          
      copying clevercsv/detect.py -> build/lib.linux-x86_64-cpython-313/clevercsv                                                                                                                                                                                                                                  
      copying clevercsv/exceptions.py -> build/lib.linux-x86_64-cpython-313/clevercsv                                                                                                                                                                                                                              
      copying clevercsv/dialect.py -> build/lib.linux-x86_64-cpython-313/clevercsv                                                                                                                                                                                                                                 
      copying clevercsv/__main__.py -> build/lib.linux-x86_64-cpython-313/clevercsv                                                                                                                                                                                                                                
      copying clevercsv/__init__.py -> build/lib.linux-x86_64-cpython-313/clevercsv                                                                                                                                                                                                                                
      copying clevercsv/break_ties.py -> build/lib.linux-x86_64-cpython-313/clevercsv                                                                                                                                                                                                                              
      copying clevercsv/dict_read_write.py -> build/lib.linux-x86_64-cpython-313/clevercsv                                                                                                                                                                                                                         
      creating build/lib.linux-x86_64-cpython-313/clevercsv/console                                                                                                                                                                                                                                                
      copying clevercsv/console/application.py -> build/lib.linux-x86_64-cpython-313/clevercsv/console                                                                                                                                                                                                             
      copying clevercsv/console/__init__.py -> build/lib.linux-x86_64-cpython-313/clevercsv/console                                                                                                                                                                                                                
      creating build/lib.linux-x86_64-cpython-313/clevercsv/console/commands                                                                                                                                
      copying clevercsv/console/commands/explore.py -> build/lib.linux-x86_64-cpython-313/clevercsv/console/commands                                                                                        
      copying clevercsv/console/commands/standardize.py -> build/lib.linux-x86_64-cpython-313/clevercsv/console/commands                                 
      copying clevercsv/console/commands/_utils.py -> build/lib.linux-x86_64-cpython-313/clevercsv/console/commands                                                                                         
      copying clevercsv/console/commands/_docs.py -> build/lib.linux-x86_64-cpython-313/clevercsv/console/commands                                                                                          
      copying clevercsv/console/commands/code.py -> build/lib.linux-x86_64-cpython-313/clevercsv/console/commands                                                                                           
      copying clevercsv/console/commands/detect.py -> build/lib.linux-x86_64-cpython-313/clevercsv/console/commands                                                                                                                                                                                                
      copying clevercsv/console/commands/__init__.py -> build/lib.linux-x86_64-cpython-313/clevercsv/console/commands
      copying clevercsv/console/commands/view.py -> build/lib.linux-x86_64-cpython-313/clevercsv/console/commands
      running egg_info
      writing clevercsv.egg-info/PKG-INFO
      writing dependency_links to clevercsv.egg-info/dependency_links.txt
      writing entry points to clevercsv.egg-info/entry_points.txt
      writing requirements to clevercsv.egg-info/requires.txt
      writing top-level names to clevercsv.egg-info/top_level.txt
      reading manifest file 'clevercsv.egg-info/SOURCES.txt'
      reading manifest template 'MANIFEST.in'
      adding license file 'LICENSE'
      writing manifest file 'clevercsv.egg-info/SOURCES.txt'
      copying clevercsv/py.typed -> build/lib.linux-x86_64-cpython-313/clevercsv
      running build_ext
      building 'clevercsv.cparser' extension
      creating build/temp.linux-x86_64-cpython-313/src
      cc -pthread -fno-strict-overflow -Wsign-compare -Wunreachable-code -DNDEBUG -g -O3 -Wall -fPIC -fPIC -I/home/brent/test_installation/uv/cache/builds-v0/.tmpeV3lcm/include -I/home/brent/test_installation/uv/python/cpython-3.13.3-linux-x86_64-gnu/include/python3.13 -c src/cparser.c -o
      build/temp.linux-x86_64-cpython-313/src/cparser.o

      [stderr]
      /home/brent/test_installation/uv/cache/builds-v0/.tmpeV3lcm/lib/python3.13/site-packages/setuptools/dist.py:759: SetuptoolsDeprecationWarning: License classifiers are deprecated.
      !!

              ********************************************************************************
              Please consider removing the following classifiers in favor of a SPDX license expression:

              License :: OSI Approved :: MIT License

              See https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#license for details.
              ********************************************************************************

      !!
        self._finalize_license_expression()
      warning: no files found matching 'requirements.txt'
      warning: no files found matching '*' under directory 'bin'
      no previously-included directories found matching 'tests/test_integration'
      warning: no previously-included files found matching 'Makefile'
      warning: no previously-included files found matching '.gitignore'
      warning: no previously-included files found matching '.travis.yml'
      warning: no previously-included files found matching '.readthedocs.yml'
      warning: no previously-included files found matching 'make_release.py'
      warning: no previously-included files found matching 'cgrep'
      warning: no previously-included files found matching 'vgrep'
      no previously-included directories found matching 'notes'
      no previously-included directories found matching 'auxiliary'
      error: command 'cc' failed: No such file or directory

      hint: This usually indicates a problem with the package or the build environment.
ERROR: Unable to continue

Please refer to Mathesar's documentation or reach out to the Mathesar team for additional support:
  * Documentation      : https://docs.mathesar.org
  * Github repo        : https://github.com/mathesar-foundation/mathesar
  * Community channels : https://mathesar.org/community
```

pavish commented on build-scripts/install/install.sh:
```diff
@@ -0,0 +1,422 @@
+install_uv() {
+  info "Installing uv..."
+  clear_uv_cache_lock
+  run_cmd /usr/bin/env bash "${INSTALL_DIR}"/uv-installer.sh
```
Fixed in https://github.com/mathesar-foundation/mathesar/pull/4399/commits/6bceea0c27f483068ea4608e9a92cfdb0573948f

mathemancer reviewed (approved):
Bash script looks okay for the most part. I found a bunch of problems with the documentation, some of which result in failed installations.

I did test that I could get Mathesar running using these instructions. I couldn't, but I noted where things broke in other comments. However, using a combination of the new instructions and my own knowledge of Mathesar led to a successful installation.

I'm going to approve this PR since a full feedback cycle takes awhile, _but_ @pavish please make sure you fix some of the documentation issues, since there's little or no point in QA of the installation process if it's not working as per the instructions.

Regarding the overall effect on the "from scratch" installation: To be frank, I'm extremely skeptical in the lack of value add from `uv` here. This installation process fails the "feels like a real binary" test for me.

- There's too much configuration beforehand to get PostgreSQL setup.
- There's too much manual config of systemd units when using this for a server (could be automated)
- The `mathesar` command is wonky, and will only be in the path for the user who installed it on the system.

I do think this is an improvement in user experience (when it works), and I think this may be the best we can do with our current stack. However, I'm even more convinced that we need to prioritize architectural and stack changes for Mathesar to further improve our installation experience.

pavish commented on docs/docs/administration/install-from-scratch.md:
````diff
@@ -1,311 +1,257 @@
 
     ```
-    chown -R gunicorn:gunicorn .media/
+    chown -R mathesar:mathesar .media/
````
Addressed in https://github.com/mathesar-foundation/mathesar/pull/4399/commits/eafaf208f9a8edb0da85423e013788c42d9c284a

pavish commented on docs/docs/administration/install-from-scratch.md:
````diff
@@ -1,311 +1,257 @@
 
     ```
-    touch /lib/systemd/system/gunicorn.service
+    touch /lib/systemd/system/mathesar.service
````
Addressed in https://github.com/mathesar-foundation/mathesar/pull/4399/commits/eafaf208f9a8edb0da85423e013788c42d9c284a

pavish commented on docs/docs/administration/install-from-scratch.md:
```diff
@@ -1,311 +1,257 @@
+1. Choose a directory where you will install Mathesar.
 
-1. Type your installation directory into the box below. Do not include a trailing slash.
+    For example: `$HOME/mathesar`, or `/etc/mathesar`, or `/opt/mathesar`.
```
Addressed in https://github.com/mathesar-foundation/mathesar/pull/4399/commits/eafaf208f9a8edb0da85423e013788c42d9c284a

pavish commented on docs/docs/administration/install-from-scratch.md:
```diff
@@ -1,311 +1,257 @@
+    RuntimeDirectory=mathesar
+    WorkingDirectory="xMATHESAR_INSTALL_DIRx"
+    ExecStart=/bin/bash -c '"xMATHESAR_INSTALL_DIRx"/bin/mathesar run'
+    EnvironmentFile=xMATHESAR_INSTALL_DIRx/.env
```
Addressed in https://github.com/mathesar-foundation/mathesar/pull/4399/commits/eafaf208f9a8edb0da85423e013788c42d9c284a

pavish commented on docs/docs/administration/install-from-scratch.md:
```diff
@@ -1,311 +1,257 @@
+    User=mathesar
+    Group=mathesar
+    RuntimeDirectory=mathesar
+    WorkingDirectory="xMATHESAR_INSTALL_DIRx"
```
Addressed in https://github.com/mathesar-foundation/mathesar/pull/4399/commits/eafaf208f9a8edb0da85423e013788c42d9c284a

pavish commented on docs/docs/administration/install-from-scratch.md:
````diff
@@ -1,311 +1,257 @@
+    ```
+
+
+### Configure Caddy
````
Addressed in https://github.com/mathesar-foundation/mathesar/pull/4399/commits/eafaf208f9a8edb0da85423e013788c42d9c284a

pavish commented on docs/docs/administration/install-from-scratch.md:
```diff
@@ -1,311 +1,257 @@
-1. Let's create a Postgres user for Mathesar
+    _(Modify as necessary based on your Postgres installation.)_
+
+1. Create a Postgres user for Mathesar.
```
Addressed in https://github.com/mathesar-foundation/mathesar/pull/4399/commits/eafaf208f9a8edb0da85423e013788c42d9c284a

pavish commented:
@mathemancer I've addressed almost all your comments, and added additional troubleshooting steps in https://github.com/mathesar-foundation/mathesar/commit/eafaf208f9a8edb0da85423e013788c42d9c284a.

I'll go ahead and merge this PR in, and we could take up any disagreements/issues, if any, in subsequent PRs.
